File: src/bin/cli/cmd/translation_status.rs
use crate::cli::OutputFormat;
use crate::dto::{ChunkDto, TranslationStatusDto};
use crate::json_output::{output_error, output_success};
use crate::utils;
use lib::subtitle::model::Subtitle;
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use lib::translation_status::service;
use log::{debug, info};

pub fn handle(
    reference: &str,
    translation: &str,
    chunk_size: usize,
    format: &OutputFormat,
) -> anyhow::Result<()> {
    info!(
        "checking translation status for {} against {}",
        translation, reference
    );

    let (ref_subs, ref_filename) = load_subtitle_file(reference, format)?;
    let (translation_subs, translation_filename) = load_subtitle_file(translation, format)?;

    debug!(
        "loaded {} reference and {} translation subtitles",
        ref_subs.len(),
        translation_subs.len()
    );

    if ref_subs.is_empty() {
        output_error(format, "empty_file", "reference file is empty", None);
        return Err(anyhow::anyhow!(""));
    }

    let report = service::check_translation_status(
        ref_subs,
        ref_filename,
        translation_subs,
        translation_filename,
        chunk_size,
    );

    let dto = TranslationStatusDto {
        ref_file: report.ref_file.clone(),
        translation_file: report.translation_file.clone(),
        total_count: report.total_count,
        translated_count: report.translated_count,
        missing_count: report.missing_count,
        progress_percentage: report.progress_percentage(),
        is_complete: report.is_complete(),
        next_chunk: report.next_chunk.as_ref().map(|c| ChunkDto {
            start_index: c.start_index,
            end_index: c.end_index,
        }),
    };

    output_success(format, &dto, || {
        display_report(&report);
    });

    info!(
        "translation progress: {}/{} ({:.1}%)",
        report.translated_count,
        report.total_count,
        report.progress_percentage()
    );

    Ok(())
}

fn load_subtitle_file(
    file: &str,
    format: &OutputFormat,
) -> anyhow::Result<(Vec<Subtitle>, String)> {
    debug!("loading subtitle file: {}", file);

    let resolved = utils::resolve_existing_path(file)?;
    let service = SubRipService::new(resolved.base_dir);

    match service.get_all(&resolved.filename) {
        Ok(subtitles) => {
            info!(
                "successfully loaded {} subtitle(s) from {}",
                subtitles.len(),
                resolved.filename
            );
            Ok((subtitles, resolved.filename))
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

fn display_report(report: &lib::translation_status::model::TranslationStatusReport) {
    let percentage = report.progress_percentage();

    if report.is_complete() {
        println!(
            "Progress: {}/{} (100%)",
            report.translated_count, report.total_count
        );
    } else {
        println!(
            "Progress: {}/{} ({:.1}%)",
            report.translated_count, report.total_count, percentage
        );
    }

    if let Some(chunk) = &report.next_chunk {
        println!("Next chunk: {}-{}", chunk.start_index, chunk.end_index);
    } else if report.is_complete() {
        println!("Translation complete!");
    }
}

File: internal/cli/cmd/translation_status.go
package cmd

import (
	"errors"
	"fmt"
	"log/slog"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/subtitle"
	"github.com/lebe-dev/submarine/pkg/translationstatus"
)

// HandleTranslationStatus is a 1-to-1 port of the Rust
// `cmd::translation_status::handle`.
//
// It loads the reference and translation subtitle files, computes the
// translation progress, and outputs the report. It returns a non-nil error
// (with an empty message, like the Rust `anyhow::anyhow!("")`) when a
// precondition fails or a file fails to load.
func HandleTranslationStatus(
	reference string,
	translation string,
	chunkSize int,
	format cli.OutputFormat,
) error {
	slog.Info(fmt.Sprintf("checking translation status for %s against %s", translation, reference))

	refSubs, refFilename, err := translationStatusLoadSubtitleFile(reference, format)
	if err != nil {
		return err
	}

	translationSubs, translationFilename, err := translationStatusLoadSubtitleFile(translation, format)
	if err != nil {
		return err
	}

	slog.Debug(fmt.Sprintf("loaded %d reference and %d translation subtitles",
		len(refSubs), len(translationSubs)))

	if len(refSubs) == 0 {
		cli.OutputError(format, "empty_file", "reference file is empty", nil)
		return errors.New("")
	}

	report := translationstatus.CheckTranslationStatus(
		refSubs,
		refFilename,
		translationSubs,
		translationFilename,
		chunkSize,
	)

	var nextChunk *cli.ChunkDto
	if report.NextChunk != nil {
		nextChunk = &cli.ChunkDto{
			StartIndex: report.NextChunk.StartIndex,
			EndIndex:   report.NextChunk.EndIndex,
		}
	}

	dto := cli.TranslationStatusDto{
		RefFile:            report.RefFile,
		TranslationFile:    report.TranslationFile,
		TotalCount:         report.TotalCount,
		TranslatedCount:    report.TranslatedCount,
		MissingCount:       report.MissingCount,
		ProgressPercentage: cli.Percentage(report.ProgressPercentage()),
		IsComplete:         report.IsComplete(),
		NextChunk:          nextChunk,
	}

	cli.OutputSuccess(format, dto, func() {
		translationStatusDisplayReport(&report)
	})

	slog.Info(fmt.Sprintf("translation progress: %d/%d (%.1f%%)",
		report.TranslatedCount, report.TotalCount, report.ProgressPercentage()))

	return nil
}

// translationStatusLoadSubtitleFile is a 1-to-1 port of the private
// `load_subtitle_file` in translation_status.rs. On error it emits a structured
// CLI error and returns an empty (anyhow) error.
func translationStatusLoadSubtitleFile(
	file string,
	format cli.OutputFormat,
) ([]subtitle.Subtitle, string, error) {
	slog.Debug(fmt.Sprintf("loading subtitle file: %s", file))

	resolved, err := cli.ResolveExistingPath(file)
	if err != nil {
		return nil, "", err
	}

	service := subtitle.NewSubRipService(resolved.BaseDir)

	subtitles, err := service.GetAll(resolved.Filename)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, file)
			cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
		}
		return nil, "", errors.New("")
	}

	slog.Info(fmt.Sprintf("successfully loaded %d subtitle(s) from %s", len(subtitles), resolved.Filename))
	return subtitles, resolved.Filename, nil
}

// translationStatusDisplayReport is a 1-to-1 port of the private
// `display_report` in translation_status.rs.
func translationStatusDisplayReport(report *translationstatus.TranslationStatusReport) {
	percentage := report.ProgressPercentage()

	if report.IsComplete() {
		fmt.Printf("Progress: %d/%d (100%%)\n", report.TranslatedCount, report.TotalCount)
	} else {
		fmt.Printf("Progress: %d/%d (%.1f%%)\n", report.TranslatedCount, report.TotalCount, percentage)
	}

	if report.NextChunk != nil {
		fmt.Printf("Next chunk: %d-%d\n", report.NextChunk.StartIndex, report.NextChunk.EndIndex)
	} else if report.IsComplete() {
		fmt.Println("Translation complete!")
	}
}
